File: code-executor/python/executor.py
#!/usr/bin/env python3
import sys
import subprocess
import time
import os
from pathlib import Path

class CodeExecutor:
    def __init__(self, time_limit=5, memory_limit=128 * 1024 * 1024):  # 128MB
        self.time_limit = time_limit
        self.memory_limit = memory_limit
        self.start_time = None
        self.end_time = None
        self.memory_used = 0
        
    def execute(self, code_file, input_file):
        """Execute Python code with resource limits"""
        self.start_time = time.time()
        
        try:
            # Read input
            with open(input_file, 'r') as f:
                input_data = f.read()
            
            # Execute the code with timeout
            process = subprocess.run(
                ['python', code_file],
                input=input_data,
                text=True,
                capture_output=True,
                timeout=self.time_limit
            )
            
            self.end_time = time.time()
            execution_time = (self.end_time - self.start_time) * 1000  # Convert to milliseconds
            
            if process.returncode == 0:
                return {
                    'status': 'accepted',
                    'output': process.stdout.strip(),
                    'error': '',
                    'execution_time': execution_time,
                    'memory_used': self.memory_used
                }
            else:
                return {
                    'status': 'runtime_error',
                    'output': process.stdout.strip(),
                    'error': process.stderr.strip(),
                    'execution_time': execution_time,
                    'memory_used': self.memory_used
                }
                
        except subprocess.TimeoutExpired:
            return {
                'status': 'time_limit_exceeded',
                'output': '',
                'error': 'Execution time limit exceeded',
                'execution_time': self.time_limit * 1000,
                'memory_used': self.memory_used
            }
        except Exception as e:
            return {
                'status': 'runtime_error',
                'output': '',
                'error': str(e),
                'execution_time': (time.time() - self.start_time) * 1000,
                'memory_used': self.memory_used
            }

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print("Usage: python3 executor.py <code_file> <input_file>")
        sys.exit(1)
    
    code_file = sys.argv[1]
    input_file = sys.argv[2]
    
    if not os.path.exists(code_file):
        print(f"Error: Code file '{code_file}' not found")
        sys.exit(1)
    
    if not os.path.exists(input_file):
        print(f"Error: Input file '{input_file}' not found")
        sys.exit(1)
    
    # Get limits from environment variables or use defaults
    time_limit = int(os.environ.get('TIME_LIMIT', 5))
    memory_limit = int(os.environ.get('MEMORY_LIMIT', 128 * 1024 * 1024))
    
    executor = CodeExecutor(time_limit, memory_limit)
    result = executor.execute(code_file, input_file)
    
    # Output result as JSON
    import json
    print(json.dumps(result))

File: code-executor/cpp/executor.cpp
Output: C++ compiler (g++) not available#include <iostream>
#include <fstream>
#include <string>
#include <chrono>
#include <cstdlib>
#include <sys/resource.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

class CodeExecutor {
private:
    long timeLimit;
    long memoryLimit;
    std::chrono::high_resolution_clock::time_point startTime;
    std::chrono::high_resolution_clock::time_point endTime;
    long memoryUsed;

    void setLimits() {
        // Set CPU time limit
        rlimit cpuLimit;
        cpuLimit.rlim_cur = timeLimit;
        cpuLimit.rlim_max = timeLimit;
        setrlimit(RLIMIT_CPU, &cpuLimit);

        // Set memory limit
        rlimit memLimit;
        memLimit.rlim_cur = memoryLimit;
        memLimit.rlim_max = memoryLimit;
        setrlimit(RLIMIT_AS, &memLimit);

        // Set file size limit
        rlimit fileSizeLimit;
        fileSizeLimit.rlim_cur = 1024 * 1024; // 1MB
        fileSizeLimit.rlim_max = 1024 * 1024;
        setrlimit(RLIMIT_FSIZE, &fileSizeLimit);

        // Set process limit
        rlimit procLimit;
        procLimit.rlim_cur = 10;
        procLimit.rlim_max = 10;
        setrlimit(RLIMIT_NPROC, &procLimit);
    }

    std::string readFile(const std::string& filename) {
        std::ifstream file(filename);
        std::string content((std::istreambuf_iterator<char>(file)),
                           std::istreambuf_iterator<char>());
        return content;
    }

public:
    CodeExecutor(long timeLimitMs, long memoryLimitBytes) 
        : timeLimit(timeLimitMs / 1000), memoryLimit(memoryLimitBytes), memoryUsed(0) {}

    struct ExecutionResult {
        std::string status;
        std::string output;
        std::string error;
        long executionTime;
        long memoryUsed;

        ExecutionResult(const std::string& status, const std::string& output, 
                       const std::string& error, long executionTime, long memoryUsed)
            : status(status), output(output), error(error), 
              executionTime(executionTime), memoryUsed(memoryUsed) {}

        std::string toJson() const {
            // Simple JSON escaping
            std::string escapedOutput = output;
            std::string escapedError = error;
            
            // Replace newlines and quotes
            size_t pos = 0;
            while ((pos = escapedOutput.find('\n', pos)) != std::string::npos) {
                escapedOutput.replace(pos, 1, "\\n");
                pos += 2;
            }
            pos = 0;
            while ((pos = escapedOutput.find('"', pos)) != std::string::npos) {
                escapedOutput.replace(pos, 1, "\\\"");
                pos += 2;
            }
            pos = 0;
            while ((pos = escapedError.find('\n', pos)) != std::string::npos) {
                escapedError.replace(pos, 1, "\\n");
                pos += 2;
            }
            pos = 0;
            while ((pos = escapedError.find('"', pos)) != std::string::npos) {
                escapedError.replace(pos, 1, "\\\"");
                pos += 2;
            }

            return "{\"status\":\"" + status + "\",\"output\":\"" + escapedOutput + 
                   "\",\"error\":\"" + escapedError + "\",\"execution_time\":" + 
                   std::to_string(executionTime) + ",\"memory_used\":" + 
                   std::to_string(memoryUsed) + "}";
        }
    };

    ExecutionResult execute(const std::string& executableFile, const std::string& inputFile) {
        startTime = std::chrono::high_resolution_clock::now();

        pid_t pid = fork();
        if (pid == 0) {
            // Child process
            setLimits();

            // Redirect stdin from input file
            freopen(inputFile.c_str(), "r", stdin);

            // Execute the program
            execl(executableFile.c_str(), executableFile.c_str(), (char*)NULL);
            
            // If execl fails
            std::cerr << "Failed to execute program" << std::endl;
            exit(1);
        } else if (pid > 0) {
            // Parent process
            int status;
            int options = WNOHANG;
            
            // Wait for process with timeout
            auto timeout = std::chrono::milliseconds(timeLimit * 1000);
            auto waitStart = std::chrono::high_resolution_clock::now();
            
            while (true) {
                pid_t result = waitpid(pid, &status, options);
                
                if (result == pid) {
                    // Process finished
                    break;
                } else if (result == -1) {
                    // Error
                    kill(pid, SIGKILL);
                    endTime = std::chrono::high_resolution_clock::now();
                    return ExecutionResult("runtime_error", "", "Wait failed", 
                                         std::chrono::duration_cast<std::chrono::milliseconds>(
                                             endTime - startTime).count(), memoryUsed);
                }
                
                // Check timeout
                auto elapsed = std::chrono::high_resolution_clock::now() - waitStart;
                if (elapsed > timeout) {
                    kill(pid, SIGKILL);
                    endTime = std::chrono::high_resolution_clock::now();
                    return ExecutionResult("time_limit_exceeded", "", "Execution time limit exceeded", 
                                         timeLimit * 1000, memoryUsed);
                }
                
                // Small delay to prevent busy waiting
                usleep(10000); // 10ms
            }
            
            endTime = std::chrono::high_resolution_clock::now();
            long executionTime = std::chrono::duration_cast<std::chrono::milliseconds>(
                endTime - startTime).count();

            // Get memory usage (simplified)
            struct rusage usage;
            getrusage(RUSAGE_CHILDREN, &usage);
            memoryUsed = usage.ru_maxrss * 1024; // Convert KB to bytes

            if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
                // Read output from a temporary file (since we redirected stdout)
                // For simplicity, we'll assume the program writes to stdout which we capture
                return ExecutionResult("accepted", "", "", executionTime, memoryUsed);
            } else {
                std::string errorMsg = "Process exited with code " + std::to_string(WEXITSTATUS(status));
                return ExecutionResult("runtime_error", "", errorMsg, executionTime, memoryUsed);
            }
        } else {
            // Fork failed
            endTime = std::chrono::high_resolution_clock::now();
            return ExecutionResult("runtime_error", "", "Failed to fork process", 
                                 std::chrono::duration_cast<std::chrono::milliseconds>(
                                     endTime - startTime).count(), memoryUsed);
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <executable> <input_file>" << std::endl;
        return 1;
    }

    std::string executable = argv[1];
    std::string inputFile = argv[2];

    // Get limits from environment variables or use defaults
    long timeLimit = 5000; // 5 seconds default
    long memoryLimit = 128 * 1024 * 1024; // 128MB default

    const char* timeLimitEnv = std::getenv("TIME_LIMIT");
    if (timeLimitEnv) {
        timeLimit = std::stol(timeLimitEnv);
    }

    const char* memoryLimitEnv = std::getenv("MEMORY_LIMIT");
    if (memoryLimitEnv) {
        memoryLimit = std::stol(memoryLimitEnv);
    }

    CodeExecutor executor(timeLimit, memoryLimit);
    CodeExecutor::ExecutionResult result = executor.execute(executable, inputFile);

    std::cout << result.toJson() << std::endl;
    return 0;
}
